src/company_cmd.cpp

```diff
@@ -592,26 +592,32 @@ void StartupCompanies()
 	_next_competitor_start = 0;
 }
 
-/** Start a new competitor company if possible. */
-static bool MaybeStartNewCompany()
+/** Start new competitor companies if possible. */
+void MaybeStartNewCompany()
 {
-	if (_networking && Company::GetNumItems() >= _settings_client.network.max_companies) return false;
+	/* Count number of existing companies. */
+	uint num_companies;
+	if (_networking) num_companies = (uint)Company::GetNumItems();
 
+	/* Count number of existing AI companies. */
+	uint num_ais = 0;
 	Company *c;
-
-	/* count number of competitors */
-	uint n = 0;
 	FOR_ALL_COMPANIES(c) {
-		if (c->is_ai) n++;
+		if (c->is_ai) num_ais++;
 	}
 
-	if (n < (uint)_settings_game.difficulty.max_no_competitors) {
-		/* Send a command to all clients to start up a new AI.
-		 * Works fine for Multiplayer and Singleplayer */
-		return DoCommandP(0, CCA_NEW_AI | INVALID_COMPANY << 16, 0, CMD_COMPANY_CTRL);
+	/* Count number of AI companies to start. */
+	uint num_to_start;
+	for (num_to_start = 0; _settings_game.difficulty.max_no_competitors - num_to_start > 0 && AI::GetStartNextTime(num_to_start) == 0; num_to_start++) {
```

Comment:
you might want to check on this statement now.
`_settings_game.difficulty.max_no_competitors - num_to_start > 0` vs `num_ais + num_to_start >= (uint)_settings_game.difficulty.max_no_competitors`. Sounds one of them is redundant.
`AI::GetStartNextTime(num_to_start) == 0` <- possibly better to put this in an `if()` statement too, with a comment to explain why you want to loop again if this is zero.

Reply:
Sorry, I reverted back to do/while cycle again, as my for cycle was bugged. I don't know how to do it, need a bit of help.
https://paste.openttdcoop.org/pixwxtrxo